### PR TITLE
Integrations for microceph rgw

### PR DESCRIPTION
**Pull request by hemanthnakkina**
# Description

Add following integrations to charm microceph
- traefik_route
- identity-service

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] CleanCode (Code refactor, test updates, does not introduce functional changes)
- [ ] Documentation update (Doc only change)

## How Has This Been Tested?

Added Unit test cases.
Manual testing with sunbeam.

## Contributor's Checklist

Please check that you have:

- [x] self-reviewed the code in this PR.
- [x] added code comments, particularly in hard-to-understand areas.
- [ ] updated the user documentation with corresponding changes.
- [x] added tests to verify effectiveness of this change.

## Depends on:
- [ ] https://github.com/canonical/microceph/pull/363

### CODE REVIEW

**Comment by hemanthnakkina** [comment]
This PR is ready for review

File: src/charm.py [hunk 1]
```diff
@@ -109,6 +112,15 @@ def _on_stop(self, event: ops.StopEvent):
             if microceph.is_cluster_member(gethostname()):
                 raise e
 
+    def _on_peer_relation_created(self, event: ops.framework.EventBase) -> None:
+        public_address = self.model.get_binding(binding_key="public").network.bind_address
```
Comment (UtkarshBhatthere) [inline src/charm.py:116]: 💯 for fetching the public bound network address!

File: src/microceph_client.py [hunk 2]
```diff
@@ -168,7 +179,26 @@ class ClusterService(BaseService):
...
+
+    def update_config(self, key: str, value: Any):
+        """Update configuration in database, create if missing."""
+        data = {"key": key, "value": value}
+        self._put("/1.0/configs", data=json.dumps(data))
+
+    def delete_config(self, key: str):
+        """Delete configuration in database if exists."""
+        data = {"key": key}
+        self._delete("/1.0/configs", data=json.dumps(data))
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply src/microceph_client.py:204]: for `delete` and `update`

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -49,13 +50,62 @@ def setUp(self):
         self.addCleanup(self.harness.cleanup)
         self.harness.begin()
 
+    def add_complete_identity_relation(self, harness: Harness) -> None:
+        """Add complete identity-service relation."""
+        credentials_content = {"username": "svcuser1", "password": "svcpass1"}
+        credentials_id = harness.add_model_secret("keystone", credentials_content)
+        app_data = {
+            "admin-domain-id": "admindomid1",
```
Comment (UtkarshBhatthere) [inline tests/unit/test_charm.py:58]: adding a note for future enhancement to support custom configuration of these values in `charm-microceph`.

**Review by UtkarshBhatthere — CHANGES_REQUESTED** [review]
This looks great to me, added one suggestion for also supporting synchronous behaviour.

File: src/microceph_client.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hemanthnakkina) [reply src/microceph_client.py:204]: Fixed

**Review by UtkarshBhatthere — DISMISSED** [review]
LGTM

File: config.yaml [hunk 4]
```diff
@@ -21,3 +21,22 @@ options:
       Supported values for this option are
       "*" - Enable RGW on all storage nodes
       ""  - Disable RGW on all storage nodes
+  region:
+    default: RegionOne
+    description: Name of the OpenStack region
+    type: string
+  namespace-tenants:
```
Comment (sabaini) [inline config.yaml:28]: Nomenclature question, would "project" not be the more modern term instead of "tenant" in OpenStack?

File: src/charm.py [hunk 5]
```diff
@@ -154,6 +174,93 @@ def channel(self, value: str) -> None:
...
             self.peers.set_app_data({"channel": value})
 
+    @property
+    def rgw_port(self):
+        """Port used for RGW service."""
+        return 80
+
+    @property
+    def service_endpoints(self):
+        """Describe the glance service endpoint."""
```
Comment (sabaini) [inline src/charm.py:184]: Probably should not refer to glance?

File: src/charm.py [hunk 6]
```diff
@@ -377,6 +570,27 @@ def configure_ceph(self, event) -> None:
                 return
             raise e
 
+    def _update_service_endpoints(self):
+        try:
+            if self.id_svc.update_service_endpoints:
+                logger.debug("Updating service endpoints after ingress relation changed")
+                self.id_svc.update_service_endpoints(self.service_endpoints)
+        except (AttributeError, KeyError):
+            pass
```
Comment (sabaini) [inline src/charm.py:579]: I'm a bit wary that we might be hiding faults in update_service_endpoints() by silently dropping an except here.
May I suggest to separate checking for existence of self.id_svc.update_service_endpoints from actual calling?

**Review by sabaini — CHANGES_REQUESTED** [review]
Hey @hemanthnakkina some small nits inline. 
It would be great if we could also get a functional test. We're installing sunbeam as part of the CI, wonder if we could do an add-relation for identity and trafik route there as well?

File: config.yaml (hunk 4 above)
Comment (hemanthnakkina) [reply config.yaml]: Done

File: src/charm.py (hunk 5 above)
Comment (hemanthnakkina) [reply src/charm.py]: Fixed

File: src/charm.py (hunk 6 above)
Comment (hemanthnakkina) [reply src/charm.py]: The attribute error can be raised if `self.id_svc.update_service_endpoints` does not exist (which should not be the case irrespective of relation joined or not) OR from `self.id_svc.update_service_endpoints(self.service_endpoints)` when the relation is not integrated.
In both the cases, the action should be ignore. So i added logging to print the error message. Any other kind of exceptions/errors in `self.id_svc.update_service_endpoints(self.service_endpoints)` results charm in blocked state.

**Comment by hemanthnakkina** [comment]
> Hey @hemanthnakkina some small nits inline. It would be great if we could also get a functional test. We're installing sunbeam as part of the CI, wonder if we could do an add-relation for identity and trafik route there as well?

@sabaini Thanks for the review. I am planning to add functional test for the integrations once the sunbeam changes are merged. However sunbeam changes will be part of 2024.1 rather than 2023.2, so i will update functional tests on next 2024.1/beta release (likely to be next week) once these PRs are merged.

**Review by sabaini — APPROVED** [review]
Thanks, lgtm!

**Pull request merged**